test/pubspec_test.dart

```diff
@@ -294,6 +297,172 @@ dependencies:
               'local pubspec.');
     });
 
+    group('source dependencies', () {
+      test('with url and name', () {
+        var pubspec = Pubspec.parse(
+          '''
+name: pkg
+dependencies:
+  foo:
+    hosted:
+      url: https://example.org/pub/
+      name: bar
+''',
+          sources,
+        );
+
+        var foo = pubspec.dependencies['foo'];
+        expect(foo.name, equals('foo'));
+        expect(foo.source, isA<HostedSource>());
```

Comment:
Let's skip this check, and leave `HostedSource` private, testing `foo.source.serializeDescription` should be fine :D